### PR TITLE
Fix parser exception in using statements with empty aliases

### LINKED ISSUE #16676: Parser throws with using statement aliases on new lines
### Prerequisites

- [X] Write a descriptive title.
- [X] Make sure you are able to repro it on the [latest released version](https://github.com/PowerShell/PowerShell/releases)
- [X] Search the existing issues.
- [X] Refer to the [FAQ](https://github.com/PowerShell/PowerShell/blob/master/docs/FAQ.md).
- [X] Refer to [Differences between Windows PowerShell 5.1 and PowerShell](https://docs.microsoft.com/powershell/scripting/whats-new/differences-from-windows-powershell).

### Steps to reproduce

Type in any of the following lines and add a linebreak after the equals sign:
```
using namespace whatever =
using module whatever =
using type whatever =
using command whatever =
```
As soon as you add the new line, PSReadline will display a parser error:

```
System.NullReferenceException: Object reference not set to an instance of an object.
   at System.Management.Automation.Language.Parser.ExtentOf(Token first, Ast last)
   at System.Management.Automation.Language.Parser.UsingStatementRule(Token usingToken)
   at System.Management.Automation.Language.Parser.UsingStatementsRule()
```
The same thing happens in PowerShell 5.1

### Expected behavior

```console
No errors thrown.
```

### Actual behavior

```console
System.NullReferenceException: Object reference not set to an instance of an object.
   at System.Management.Automation.Language.Parser.ExtentOf(Token first, Ast last)
   at System.Management.Automation.Language.Parser.UsingStatementRule(Token usingToken)
   at System.Management.Automation.Language.Parser.UsingStatementsRule()
```
```


### Error details

```console
PS C:\Users\Martin> get-error

Exception             :
    Type           : System.Management.Automation.MethodInvocationException
    ErrorRecord    :
        Exception             :
            Type    : System.Management.Automation.ParentContainsErrorRecordException
            Message : Exception calling "ReadLine" with "3" argument(s): "Unrecoverable error in PowerShell."
            HResult : -2146233087
        CategoryInfo          : NotSpecified: (:) [], ParentContainsErrorRecordException
        FullyQualifiedErrorId : ParseException
        InvocationInfo        :
            ScriptLineNumber : 7
            OffsetInLine     : 5
            HistoryId        : -1
            ScriptName       : C:\Users\Martin\Documents\PowerShell\Modules\PSReadLine\2.2.0\PSReadLine.psm1
            Line             : [Microsoft.PowerShell.PSConsoleReadLine]::ReadLine($host.Runspace, $ExecutionContext, $lastRunStatus)

            PositionMessage  : At C:\Users\Martin\Documents\PowerShell\Modules\PSReadLine\2.2.0\PSReadLine.psm1:7 char:5
                               +     [Microsoft.PowerShell.PSConsoleReadLine]::ReadLine($host.Runspace …
                               +     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
            PSScriptRoot     : C:\Users\Martin\Documents\PowerShell\Modules\PSReadLine\2.2.0
            PSCommandPath    : C:\Users\Martin\Documents\PowerShell\Modules\PSReadLine\2.2.0\PSReadLine.psm1
            CommandOrigin    : Internal
        ScriptStackTrace      : at PSConsoleHostReadLine, C:\Users\Martin\Documents\PowerShell\Modules\PSReadLine\2.2.0\PSReadLine.psm1: line 7
    TargetSite     :
        Name          : ConvertToMethodInvocationException
        DeclaringType : System.Management.Automation.ExceptionHandlingOps, System.Management.Automation, Version=7.2.0.10, Culture=neutral, PublicKeyToken=31bf3856ad364e35
        MemberType    : Method
        Module        : System.Management.Automation.dll
    Message        : Exception calling "ReadLine" with "3" argument(s): "Unrecoverable error in PowerShell."
    Data           : System.Collections.ListDictionaryInternal
    InnerException :
        Type           : System.Management.Automation.ParseException
        Message        : Unrecoverable error in PowerShell.
        ErrorRecord    :
            Exception             :
                Type    : System.Management.Automation.ParentContainsErrorRecordException
                Message : Unrecoverable error in PowerShell.
                HResult : -2146233087
            CategoryInfo          : ParserError: (:) [], ParentContainsErrorRecordException
            FullyQualifiedErrorId : Parse
        TargetSite     :
            Name          : ParseInput
            DeclaringType : System.Management.Automation.Language.Parser
            MemberType    : Method
            Module        : System.Management.Automation.dll
        InnerException :
            Type       : System.NullReferenceException
            TargetSite :
                Name          : ExtentOf
                DeclaringType : System.Management.Automation.Language.Parser
                MemberType    : Method
                Module        : System.Management.Automation.dll
            Message    : Object reference not set to an instance of an object.
            Source     : System.Management.Automation
            HResult    : -2147467261
            StackTrace :
   at System.Management.Automation.Language.Parser.ExtentOf(Token first, Ast last)
   at System.Management.Automation.Language.Parser.UsingStatementRule(Token usingToken)
   at System.Management.Automation.Language.Parser.UsingStatementsRule()
   at System.Management.Automation.Language.Parser.ScriptBlockRule(Token lCurly, Boolean isFilter, StatementAst predefinedStatementAst)
   at System.Management.Automation.Language.Parser.ParseTask(String fileName, String input, List`1 tokenList, Boolean recursed, ParseMode parseMode)
   at System.Management.Automation.Language.Parser.Parse(String fileName, String input, List`1 tokenList, ParseError[]& errors, ParseMode parseMode)
   at System.Management.Automation.Language.Parser.ParseInput(String input, String fileName, Token[]& tokens, ParseError[]& errors)
        Source         : System.Management.Automation
        HResult        : -2146233087
        StackTrace     :
   at System.Management.Automation.Language.Parser.ParseInput(String input, String fileName, Token[]& tokens, ParseError[]& errors)
   at System.Management.Automation.Language.Parser.ParseInput(String input, Token[]& tokens, ParseError[]& errors)
   at Microsoft.PowerShell.PSConsoleReadLine.ParseInput()
   at Microsoft.PowerShell.PSConsoleReadLine.GenerateRender(String defaultColor)
   at Microsoft.PowerShell.PSConsoleReadLine.ForceRender()
   at Microsoft.PowerShell.PSConsoleReadLine.Render()
   at Microsoft.PowerShell.PSConsoleReadLine.Insert(String s)
   at Microsoft.PowerShell.PSConsoleReadLine.ReadLine(Runspace runspace, EngineIntrinsics engineIntrinsics, CancellationToken cancellationToken, Nullable`1 lastRunStatus)
   at Microsoft.PowerShell.PSConsoleReadLine.ReadLine(Runspace runspace, EngineIntrinsics engineIntrinsics, Nullable`1 lastRunStatus)
   at CallSite.Target(Closure , CallSite , Type , Object , Object , Object )
    Source         : System.Management.Automation
    HResult        : -2146233087
    StackTrace     :
   at System.Management.Automation.ExceptionHandlingOps.ConvertToMethodInvocationException(Exception exception, Type typeToThrow, String methodName, Int32 numArgs, MemberInfo memberInfo)
   at CallSite.Target(Closure , CallSite , Type , Object , Object , Object )
   at System.Dynamic.UpdateDelegates.UpdateAndExecute4[T0,T1,T2,T3,TRet](CallSite site, T0 arg0, T1 arg1, T2 arg2, T3 arg3)
   at System.Management.Automation.Interpreter.DynamicInstruction`5.Run(InterpretedFrame frame)
   at System.Management.Automation.Interpreter.EnterTryCatchFinallyInstruction.Run(InterpretedFrame frame)
CategoryInfo          : NotSpecified: (:) [], MethodInvocationException
FullyQualifiedErrorId : ParseException
InvocationInfo        :
    ScriptLineNumber : 7
    OffsetInLine     : 5
    HistoryId        : -1
    ScriptName       : C:\Users\Martin\Documents\PowerShell\Modules\PSReadLine\2.2.0\PSReadLine.psm1
    Line             : [Microsoft.PowerShell.PSConsoleReadLine]::ReadLine($host.Runspace, $ExecutionContext, $lastRunStatus)

    PositionMessage  : At C:\Users\Martin\Documents\PowerShell\Modules\PSReadLine\2.2.0\PSReadLine.psm1:7 char:5
                       +     [Microsoft.PowerShell.PSConsoleReadLine]::ReadLine($host.Runspace …
                       +     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    PSScriptRoot     : C:\Users\Martin\Documents\PowerShell\Modules\PSReadLine\2.2.0
    PSCommandPath    : C:\Users\Martin\Documents\PowerShell\Modules\PSReadLine\2.2.0\PSReadLine.psm1
    CommandOrigin    : Internal
ScriptStackTrace      : at PSConsoleHostReadLine, C:\Users\Martin\Documents\PowerShell\Modules\PSReadLine\2.2.0\PSReadLine.psm1: line 7
```

### Environment data

```powershell
PS C:\Users\Martin> $PSVersionTable

Name                           Value
----                           -----
PSVersion                      7.2.0-preview.10
PSEdition                      Core
GitCommitId                    7.2.0-preview.10
OS                             Microsoft Windows 10.0.19044
Platform                       Win32NT
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0…}
PSRemotingProtocolVersion      2.3
SerializationVersion           1.1.0.1
WSManStackVersion              3.0
```

### Visuals

_No response_

### PR DESCRIPTION
**Pull request by MartinGC94**
# PR Summary
Fixes https://github.com/PowerShell/PowerShell/issues/16676

The parser threw an exception when entering a new line after the equals token, and when you entered a comma after the equals sign like in these 2 examples:
```
using namespace whatever =

using namespace whatever = ,
```

## PR Checklist

- [X] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [X] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [X] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [X] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [X] None
    - **OR**
    - [ ] [Experimental feature(s) needed](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/staging/reference/6/Microsoft.PowerShell.Core/About/about_Experimental_Features.md)
        - [ ] Experimental feature name(s): 
- **User-facing changes**
    - [X] Not Applicable
    - **OR**
    - [ ] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed: 
- **Testing - New and feature**
    - [ ] N/A or can only be tested interactively
    - **OR**
    - [X] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
- **Tooling**
    - [X] I have considered the user experience from a tooling perspective and don't believe tooling will be impacted.
    - **OR**
    - [ ] I have considered the user experience from a tooling perspective and opened an issue in the relevant tool repository. This may include:
        - [ ] Impact on [PowerShell Editor Services](https://github.com/PowerShell/PowerShellEditorServices) which is used in the [PowerShell extension](https://github.com/PowerShell/vscode-powershell) for VSCode
        (which runs in a different PS Host).
            - [ ] Issue filed: 
        - [ ] Impact on Completions (both in the console and in editors) - one of PowerShell's most powerful features.
            - [ ] Issue filed: 
        - [ ] Impact on [PSScriptAnalyzer](https://github.com/PowerShell/PSScriptAnalyzer) (which provides linting & formatting in the editor extensions).
            - [ ] Issue filed: 
        - [ ] Impact on [EditorSyntax](https://github.com/PowerShell/EditorSyntax) (which provides syntax highlighting with in VSCode, GitHub, and many other editors).
            - [ ] Issue filed:

### CODE REVIEW

File: test/powershell/Language/Parser/Parsing.Tests.ps1 [hunk 1]
```diff
@@ -644,3 +644,19 @@ Describe "Parsing array that has too many dimensions" -Tag CI {
         }
     }
 }
+
+Describe "Parsing using statement with alias and linebreak" -Tag CI {
+    It "ParseError for '<Script>'" -TestCases @(
+        @{ Script = "using namespace x =`n"; ErrorId = @('MissingNamespaceAlias'); StartOffset = @(19); EndOffset = @(19) }
```
Comment (iSazonov) [inline test/powershell/Language/Parser/Parsing.Tests.ps1:650]: Will this work if there are spaces after equal sign? "using namespace x =    `n"

Comment (MartinGC94) [reply test/powershell/Language/Parser/Parsing.Tests.ps1:650]: Yes, whitespaces are skipped over so the code only needs to explicitly handle "real" tokens.

Comment (iSazonov) [reply test/powershell/Language/Parser/Parsing.Tests.ps1:650]: Nice! Please add a test with space(s) - since it is _Parser_ we should consider edge cases to avoid regressions.

**Review by vexx32 — APPROVED** [review]

**Comment by iSazonov** [comment]
@andschwa Have you any concerns about PSES?

**Review by SeeminglyScience — APPROVED** [review]
LGTM! No PSES concerns I can think of.

**Review by iSazonov — APPROVED** [review]

**Comment by andschwa** [comment]
I agree with @SeeminglyScience, I don't see an impact on PSES.

**Comment by MartinGC94** [comment]
I also discovered that commas have the same problem: `using command X = ,`  causes an error.

File: src/System.Management.Automation/engine/parser/Parser.cs [hunk 2]
```diff
@@ -5010,14 +5010,20 @@ private StatementAst UsingStatementRule(Token usingToken)
                     SkipToken();
 
                     var aliasToken = NextToken();
-                    if (aliasToken.Kind == TokenKind.EndOfInput)
+                    if (aliasToken.Kind is TokenKind.EndOfInput or TokenKind.NewLine or TokenKind.Semi)
```
Comment (iSazonov) [inline src/System.Management.Automation/engine/parser/Parser.cs:5013]: `or`?

Comment (MartinGC94) [reply src/System.Management.Automation/engine/parser/Parser.cs:5013]: https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/operators/patterns#logical-patterns it should work the same as the longer version: `aliasToken.Kind == TokenKind.EndOfInput || aliasToken.Kind == TokenKind.NewLine || aliasToken.Kind == TokenKind.Semi`  
If you are asking why I'm checking for EndOfInput + the others then it's because without them the parser throws an error in certain conditions.

**Comment by iSazonov** [comment]
@MartinGC94 Please look test error.

File: src/System.Management.Automation/engine/parser/Parser.cs [hunk 3]
```diff
@@ -5010,14 +5010,19 @@ private StatementAst UsingStatementRule(Token usingToken)
...
                     {
                         UngetToken(aliasToken);
                         ReportIncompleteInput(After(equalsToken),
                             nameof(ParserStrings.MissingNamespaceAlias),
                             ParserStrings.MissingNamespaceAlias);
                         return new ErrorStatementAst(ExtentOf(usingToken, equalsToken));
                     }
+                    if (aliasToken.Kind == TokenKind.Comma)
```
Comment (iSazonov) [inline src/System.Management.Automation/engine/parser/Parser.cs:5021]: Please add blank line.
```suggestion

                    if (aliasToken.Kind == TokenKind.Comma)
```

**Comment by iSazonov** [comment]
@MartinGC94 I reviewed all code paths with `GetCommandArgument()` and found the same issue if we type `configuration ,`. 
It is in `private ExpressionAst GetWordOrExpression(Token keywordToken)`. There is `GetCommandArgument()` call with only `TokenKind.EndOfInput` check.

It would be great if you fixed this too.

File: src/System.Management.Automation/engine/parser/Parser.cs [hunk 4]
```diff
@@ -5019,14 +5019,31 @@ private StatementAst UsingStatementRule(Token usingToken)
...
+                        Ast[] nestedAsts;
+                        if (aliasAst is null)
+                        {
+                            nestedAsts = new Ast[] { itemAst };
+                        }
+                        else
+                        {
+                            nestedAsts = new Ast[] { itemAst, aliasAst };
+                        }
+                        ReportError(errorExtent, nameof(ParserStrings.InvalidValueForUsingItemName), ParserStrings.InvalidValueForUsingItemName, errorExtent.Text);
```
Comment (iSazonov) [inline src/System.Management.Automation/engine/parser/Parser.cs:5045]:
```suggestion

                        ReportError(errorExtent, nameof(ParserStrings.InvalidValueForUsingItemName), ParserStrings.InvalidValueForUsingItemName, errorExtent.Text);
```

**Comment by MartinGC94** [comment]
I will take a look later, but typing `configuration ,` into the console doesn't cause an exception on my system so that scenario may already be handled elsewhere.

**Comment by iSazonov** [comment]
Assert throw in debug build
image

**Review by iSazonov — APPROVED** [review]

**Comment by adityapatwardhan** [comment]
@MartinGC94 Thank you for your contribution!
